src/nvim/api/buffer.c

```diff
@@ -170,20 +192,34 @@ void buffer_set_line_slice(Buffer buffer,
                       ArrayOf(String) replacement,
                       Error *err)
 {
+  if (start < 0) start -= 1;
+  if (end < 0) end -= 1;
```

Comment:
This and the other cases should use `normalize_index` or a new similar private function.